pkg/platform/authentication/auth.go

```diff
@@ -113,9 +116,51 @@ func (s *Auth) Close() error {
 	return nil
 }
 
+func (s *Auth) storeAuthenticatedDevice(deviceCode strfmt.UUID, accessToken *mono_models.JWT) error {
+	defer s.updateRollbarPerson()
+
+	s.user = accessToken.User
+	s.bearerToken = accessToken.Token
+	clientAuth := httptransport.BearerToken(s.bearerToken)
+	s.clientAuth = &clientAuth
+
+	err := s.cfg.Set(deviceCodeConfigKey, deviceCode)
+	if err != nil {
+		return errs.Wrap(err, "Could not set deviceCode credentials in config")
+	}
+
+	return err
+}
+
 // Authenticated checks whether we are currently authenticated
 func (s *Auth) Authenticated() bool {
-	return s.clientAuth != nil
+	if s.clientAuth != nil {
+		return true
+	}
+	existingDeviceCode := s.cfg.GetString(deviceCodeConfigKey)
+	if existingDeviceCode == "" || s.bearerToken != "" {
+		return false
+	}
+	// Check if the device is still authenticated with the Platform and if so, get a token.
+	deviceCode := strfmt.UUID(existingDeviceCode)
+	params := oauth.NewAuthDeviceGetParams()
+	params.SetDeviceCode(deviceCode)
+	if response, err := mono.Get().Oauth.AuthDeviceGet(params); response != nil {
+		if err := s.storeAuthenticatedDevice(deviceCode, response.Payload.AccessToken); err == nil {
+			return true
+		}
+		if err != nil {
+			logging.Error("Error storing authenticated device", err.Error())
+		}
+	} else {
+		// Either the token for deviceCode has expired, we are rate-limited by the Platform and
+		// have to try again later, or the Platform is unreachable.
+		// Rate limiting can happen during testing.
+		if badRequest, ok := err.(*oauth.AuthDeviceGetBadRequest); ok && *badRequest.Payload.Error == oauth.AuthDeviceGetBadRequestBodyErrorSlowDown {
+			logging.Warning("Attempting to query the Platform for device authentication status too frequently.")
+		}
+	}
+	return false
```

Comment:
None of this should be required if we use the API key to store the authenticated state.

Reply:
Agreed.